src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -2691,7 +2691,7 @@ public void ReadAsyncCallback(IntPtr key, PacketHandle packet, uint error)
 
             // The mars physical connection can get a callback
             // with a packet but no result after the connection is closed.
-            if (source == null && _parser._pMarsPhysicalConObj == this)
```

Comment:
This condition removal doesn't seem right. This shouldn't be done in Native SNI. In native SNI all callbacks happen on the same thread and this differentiation of the physical connection object is the only way to differentiate if the callback corresponds to this object.

Also why is this correct for Managed SNI?

Reply:
Because I've introduced the close method on disconnection the physical connection will now be nulled, so checking if it's the current instance doesn't help. This catches the case where the socket is closed while the socket wait is in progress which causes a 0 packet receive.

The socket in the native version is handled in the sni lib and the same condition doesn't occur there.

Reply:
> Because I've introduced the close method on disconnection the physical connection will now be nulled

What close method?

Reply:
In SNIMarsHadle.HandleReceiveError, 
```csharp
if (sniErrorCode == TdsEnums.SNI_WSAECONNRESET)
{
    TdsParser parser = _callbackObject.Parser;
    parser.State = TdsParserState.Broken;
    parser.Connection.BreakConnection();
}
```
BreakConnection dooms closes and does a ton of other stuff one of which is nulling the state object. Since the only two values that object can have are null and this and if it's either null or this and the source tcs is null it's not going to work the check on the state object seemed redundant.

Reply:
The BreakConnection, only impacts breaking the connection for Managed SNI, this code would never run for Native SNI, and in that case the removal of this check can have unpredictable impact when Native SNI is being used.

Reply:
I see. Then I'm going to have to put that back in and do some debugging and tracing to repo the problem it fixes and see if there's another way to detect it strictly in the managed side. I think if i'd seen one I would have used it so this might be tricky.